File: Dijkstra/Python/dijkstra.py
import heapq
import math
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.animation as animation

class Dijkstra:
    def __init__(self,grid,start,goal,radius,board_size):
        self.grid = grid
        self.start = start
        self.goal = goal
        self.radius = radius
        self.board_size = board_size

    class Node:
        def __init__(self,directions,cost):
            self.directions = directions
            self.cost = cost
    
    def get_neighbors(self, node):
        neighbors = []
        distances = []
        nexts = [self.Node((-1, 0),1), self.Node((0, 1),1), self.Node((0, -1),1), self.Node((1,0),1),
                self.Node((-1,1),math.sqrt(2)), self.Node((1,1),math.sqrt(2)),self.Node((1, -1),math.sqrt(2)), self.Node((-1,-1),math.sqrt(2))]
        for next in nexts:
            neighbor = (node[0] + next.directions[0], node[1] + next.directions[1])
            if self.board_size <= neighbor[0] < len(self.grid)-self.board_size and self.board_size <= neighbor[1] < len(self.grid[0])-self.board_size:
                if self.grid[neighbor[0]][neighbor[1]] == 0:
                    if next.directions[0] != 0 and next.directions[1] != 0:  # 对角线方向
                        if self.grid[node[0] + next.directions[0]][node[1]] == 0 and self.grid[node[0]][node[1] + next.directions[1]] == 0:
                            neighbors.append(neighbor)
                            distances.append(next.cost)
                    else:
                        neighbors.append(neighbor)
                        distances.append(next.cost)

        return neighbors,distances

    def plan(self):
        priority_queue = []
        heapq.heappush(priority_queue,(0,self.start))
        costs = {self.start: 0}
        previous_nodes = {self.start: None}
        self.visited = set()
        self.visit_order = []

        while priority_queue:
            current_cost, current_node = heapq.heappop(priority_queue)
            # Determines whether the current node has already been visited
            if current_node in self.visited:
                continue

            self.visited.add(current_node)
            self.visit_order.append(current_node)
            if current_node == self.goal:
                break
            # Find passable neighbors
            neighbors, distances = self.get_neighbors(current_node)
            for neighbor,distance in zip(neighbors,distances):
                if neighbor[0] == -1:
                    continue
                # Compute the cost from the start point
                cost = current_cost + distance
                # Store cost and update to minimum value
                if neighbor not in costs or cost < costs[neighbor]:
                    costs[neighbor] = cost
                    # The parent node of neighbor is current_node
                    previous_nodes[neighbor] = current_node
                    # Push the node into priority_queue
                    heapq.heappush(priority_queue,(cost, neighbor))
        
        self.path = []
        current_node = self.goal
        while current_node is not None:
            self.path.append(current_node)
            current_node = previous_nodes.get(current_node)
        self.path = self.path[::-1]

        return costs[self.goal]

    def visualize_grid(self, cost):
        fig, ax = plt.subplots()
        grid = np.array(self.grid)

        plt.imshow(grid, cmap='Greys', origin='upper')

        # Mark the start and goal point
        plt.scatter(self.start[1], self.start[0], c='green', marker='o', s=60)
        plt.scatter(self.goal[1], self.goal[0], c='blue', marker='o', s=60)

        # # Mark locations which has been visited
        # for node in self.visited:
        #     plt.gca().add_patch(plt.Rectangle((node[1]-0.5, node[0]-0.5), 1, 1, fill=True, color='gray', alpha=0.5))
            
        # # 标记路径
        # if self.path:
        #     path_x, path_y = zip(*self.path)
        #     plt.plot(path_y, path_x, c='red', linewidth=2)

        visited_patches = []
        path_line, = ax.plot([], [], c='red', linewidth=2)
        for order in range(len(self.visit_order)):
            node = self.visit_order[order]
            plt.Rectangle((node[1]-0.5, node[0]-0.5), 1, 1, fill=True, color='gray', alpha=0.5)

        def update(frame):
            if frame < len(self.visit_order):
                node = self.visit_order[frame]
                patch = plt.Rectangle((node[1]-0.5, node[0]-0.5), 1, 1, fill=True, color='gray', alpha=0.5)
                visited_patches.append(patch)
                ax.add_patch(patch)
            else:
                path_x, path_y = zip(*self.path)
                path_line.set_data(path_y, path_x)

            return visited_patches + [path_line]*10
    
        plt.title("Dijkstra\n"+"Cost: "+str(cost))
        ani = animation.FuncAnimation(fig, update, frames=len(self.visit_order) + 10, interval=10, repeat=False)
        ani.save("map_animate.mp4",writer='ffmpeg')

        plt.savefig("map_animate.png",bbox_inches='tight')
        plt.show()

File: Dijkstra/CPP/src/dijkstra.cpp
#include "dijkstra.hpp"
#include <vector>
#include <queue>
#include <cmath>
#include <set>
#include <map>
#include <tuple>
#include <algorithm>
#include <opencv2/opencv.hpp>

using namespace std;
using namespace cv;

std::tuple<vector<pair<int, int>>, vector<double>> Dijkstra::get_neighbors(pair<int, int> node) {
    vector<pair<int, int>> neighbors;
    vector<double> distances;
    vector<PlanNode> nexts = {PlanNode({-1, 0}, 1), PlanNode({-1, 1}, sqrt(2)), PlanNode({0, 1}, 1), PlanNode({1, 1}, sqrt(2)),
                          PlanNode({1, 0}, 1), PlanNode({1, -1}, sqrt(2)), PlanNode({0, -1}, 1), PlanNode({-1, -1}, sqrt(2))};
    for (auto& next : nexts) {
        pair<int, int> neighbor = {node.first + next.directions.first, node.second + next.directions.second};
        if (0 <= neighbor.first && neighbor.first < grid.size() && 0 <= neighbor.second && neighbor.second < grid[0].size()) {
            if (grid[neighbor.first][neighbor.second] == 0) {
                if (next.directions.first != 0 && next.directions.second != 0) {
                    if (grid[node.first + next.directions.first][node.second] == 0 && grid[node.first][node.second + next.directions.second] == 0) {
                        neighbors.push_back(neighbor);
                        distances.push_back(next.cost);
                    }
                } else {
                    neighbors.push_back(neighbor);
                    distances.push_back(next.cost);
                }
            }
        }
    }
    return make_tuple(neighbors,distances);
}

double Dijkstra::plan() {
    priority_queue<pair<double, pair<int, int>>, vector<pair<double, pair<int, int>>>, greater<>> priority_queue;
    priority_queue.push({0, start});
    map<pair<int, int>, double> costs;
    map<pair<int, int>, pair<int, int>> previous_nodes;
    costs[start] = 0;

    while (!priority_queue.empty()) {
        auto [current_cost, current_node] = priority_queue.top();
        priority_queue.pop();

        if (visited.find(current_node) != visited.end()) continue;

        visited.insert(current_node);
        visit_order.push_back(current_node);

        if (current_node == goal) break;

        vector<pair<int, int>> neighbors;
        vector<double> distances;
        tie(neighbors,distances) = get_neighbors(current_node);

        for (size_t i=0; i<neighbors.size();++i) {
            pair<int, int> neighbor = neighbors[i];
            double distance = distances[i];
            double cost = current_cost + distance;
            // double cost = current_cost + sqrt(pow(neighbor.first - current_node.first, 2) + pow(neighbor.second - current_node.second, 2));
            if (costs.find(neighbor) == costs.end() || cost < costs[neighbor]) {
                costs[neighbor] = cost;
                previous_nodes[neighbor] = current_node;
                priority_queue.push({cost, neighbor});
            }
        }
    }

    path.clear();
    pair<int, int> current_node = goal;
    while (current_node != start) {
        path.push_back(current_node);
        current_node = previous_nodes[current_node];
        std::cout<<"node: "<<current_node.first<<","<<current_node.second<<" ";
        printf("cost:%lf\n",costs[current_node]);
    }
    path.push_back(start);
    reverse(path.begin(), path.end());

    return costs[goal];
}

void Dijkstra::visualize_grid(double cost) {
    int rows = grid.size();
    int cols = grid[0].size();
    
    // Set the size of grid
    int cell_size = 20;
    int img_width = cols * cell_size;
    int img_height = rows * cell_size;

    // Create video file
    // AVI Pattern
    // VideoWriter video("dijkstra_animation.avi", VideoWriter::fourcc('M', 'J', 'P', 'G'), 100, Size(img_width, img_height));
    // MP4 Pattern
    VideoWriter video("dijkstra_animation.mp4", VideoWriter::fourcc('m', 'p', '4', 'v'), 100, Size(img_width, img_height));
    
    // Create image for each frame
    Mat img(img_height, img_width, CV_8UC3, Scalar(255, 255, 255));
    // Title and Cost
    std::string title = "Dijkstra";
    putText(img, title, Point(550, 40), FONT_HERSHEY_SIMPLEX, 0.8, Scalar(0, 0, 0), 2);
    std::string cost_text = "Cost: " + std::to_string(cost);
    putText(img, cost_text, Point(500, 80), FONT_HERSHEY_SIMPLEX, 0.8, Scalar(0, 0, 0), 2);

    // Draw Obstacles
    for (int i = 0; i < rows; ++i) {
        for (int j = 0; j < cols; ++j) {
            if (grid[i][j] == 1) {
                rectangle(img, Point(j * cell_size, i * cell_size),
                          Point((j + 1) * cell_size, (i + 1) * cell_size),
                          Scalar(0, 0, 0), FILLED);
            }
        }
    }

    // Draw start and goal point
    rectangle(img, Point(start.second * cell_size, start.first * cell_size),
            Point((start.second + 1) * cell_size, (start.first + 1) * cell_size),
            Scalar(0, 255, 0), FILLED);

    rectangle(img, Point(goal.second * cell_size, goal.first * cell_size),
            Point((goal.second + 1) * cell_size, (goal.first + 1) * cell_size),
            Scalar(0, 255, 255), FILLED);

    // Draw visited areas
    for (size_t i=1;i<visit_order.size()-1;++i) {
        const auto& node = visit_order[i];
        rectangle(img, Point(node.second * cell_size, node.first * cell_size),
        Point((node.second + 1) * cell_size, (node.first + 1) * cell_size),
        Scalar(128, 128, 128), FILLED);
        
        video.write(img);
    }

    // Draw path
    for (size_t i = 1; i < path.size(); ++i) {
        Point p1(path[i - 1].second * cell_size + cell_size / 2,
                path[i - 1].first * cell_size + cell_size / 2);
        Point p2(path[i].second * cell_size + cell_size / 2,
                path[i].first * cell_size + cell_size / 2);
        line(img, p1, p2, Scalar(0, 0, 255), 2);

        video.write(img);
    }

    video.release();
    imwrite("dijkstra_result.png", img);
    imshow("Dijkstra", img);
    waitKey(0);
    destroyAllWindows();
}